File: internal/run/depth.go
package run

import (
	"fmt"
	"os"
	"strconv"
)

// The depth is carried in the environment of every process bolt spawns, so it
// survives reparenting, backgrounding, and a task command that invokes bolt
// directly rather than through a jig task.
const (
	// DepthVariable holds how many invocations deep the spawning bolt was.
	DepthVariable = "BOLT_DEPTH"
	// CeilingVariable holds the limit in force. Bolt exports the resolved
	// ceiling rather than reading a fresh one at each level, which is what
	// stops a jig raising the limit it is running under.
	CeilingVariable = "BOLT_MAX_DEPTH"
	// DefaultCeiling is a guard against accident and runaway.
	DefaultCeiling = 4
)

// resolveDepth says how deep this invocation is and what limit it runs under.
//
// The ceiling is read from the environment only where the depth variable is
// absent, which is the outermost invocation and the only one entitled to set
// it. Every level below takes the resolved value its parent exported.
func resolveDepth(options Options) (depth, ceiling int) {
	if options.Depth > 0 {
		// Nested in process. The parent passed both down.
		return options.Depth, options.Ceiling
	}

	if raw, set := os.LookupEnv(DepthVariable); set {
		// A process bolt spawned, which is not outermost however it was
		// reached. FR-5.6 contemplates a task command invoking bolt directly.
		parent, err := strconv.Atoi(raw)
		if err != nil {
			parent = 0
		}
		return parent + 1, ceilingFromEnvironment()
	}

	return 0, ceilingFromEnvironment()
}

func ceilingFromEnvironment() int {
	raw, set := os.LookupEnv(CeilingVariable)
	if !set {
		return DefaultCeiling
	}
	value, err := strconv.Atoi(raw)
	if err != nil || value < 1 {
		return DefaultCeiling
	}
	return value
}

// environment is what every process this run spawns inherits, carrying the
// depth and the ceiling so a bolt reached any way at all knows both.
func environment(depth, ceiling int) []string {
	out := make([]string, 0, len(os.Environ())+2)
	for _, entry := range os.Environ() {
		if hasPrefix(entry, DepthVariable+"=") || hasPrefix(entry, CeilingVariable+"=") {
			continue
		}
		out = append(out, entry)
	}
	return append(out,
		fmt.Sprintf("%s=%d", DepthVariable, depth),
		fmt.Sprintf("%s=%d", CeilingVariable, ceiling),
	)
}

func hasPrefix(value, prefix string) bool {
	return len(value) >= len(prefix) && value[:len(prefix)] == prefix
}

// tooDeep says whether this invocation has passed the ceiling.
//
// The guard is against accident and runaway, not against a jig trying to defeat
// it. FR-5.6 contemplates a task command invoking bolt directly, and such a
// command can unset the variable and be believed outermost. Closing that needs
// the ancestry cross-check, which is a question rather than a row.
func tooDeep(depth, ceiling int) error {
	if depth <= ceiling {
		return nil
	}
	return fmt.Errorf("nesting reached depth %d, past the ceiling of %d", depth, ceiling)
}

File: src/depth.rs
//! How deep a run is nested, and how deep it may go.
//!
//! FR-5.6 carries the depth in the environment of **every** process bolt spawns
//! rather than passing it to child jigs alone. That is what makes it survive
//! reparenting, backgrounding, and a task command that invokes bolt directly
//! instead of through a jig task, which is the case FR-5.7a builds on and the
//! case this can be tested against before nested jigs exist.

use std::env;

/// The variable carrying the current depth, by FR-5.6a.
pub const DEPTH: &str = "BOLT_DEPTH";

/// The variable carrying the ceiling, by FR-5.6a.
pub const CEILING: &str = "BOLT_MAX_DEPTH";

/// How deep a run may nest before it is refused, by FR-5.7.
pub const DEFAULT_CEILING: u32 = 4;

/// Where a run sits, and where the floor is.
#[derive(Debug, Clone, Copy, PartialEq, Eq)]
pub struct Depth {
    /// This run's own depth, the outermost being 1.
    pub level: u32,
    /// The deepest level allowed.
    pub ceiling: u32,
}

impl Depth {
    /// Read this run's depth from the environment it was started in.
    ///
    /// The outermost invocation finds nothing set and is depth 1. Anything
    /// deeper finds what its parent exported and increments it, by FR-5.6.
    ///
    /// **A value that will not parse is treated as absent**, which reads the run
    /// as outermost rather than refusing it. A caller's environment is not a
    /// document bolt was asked to validate, and FR-5.7a already says the ceiling
    /// guards against accident rather than evasion, so refusing here would fail
    /// runs over stray shell state while stopping nobody who meant it.
    #[must_use]
    pub fn from_environment() -> Self {
        let inherited = read(DEPTH);
        Self {
            level: inherited.map_or(1, |level| level.saturating_add(1)),
            ceiling: read(CEILING).unwrap_or(DEFAULT_CEILING),
        }
    }

    /// Whether this run is deeper than it is allowed to be, by FR-5.7.
    #[must_use]
    pub fn exceeded(self) -> bool {
        self.level > self.ceiling
    }

    /// What every process this run spawns is told, by FR-5.6.
    ///
    /// FR-5.7 says the ceiling is read from the environment only at the
    /// outermost invocation, and **this is the mechanism rather than a check**:
    /// bolt overwrites both variables on every spawn, so a nested bolt reading
    /// them gets what the bolt above it set and never what a jig wrote. There
    /// is no branch on being outermost, and adding one would only matter for a
    /// command deliberately rewriting the variable, which FR-5.7a puts out of
    /// scope.
    #[must_use]
    pub fn exported(self) -> [(&'static str, String); 2] {
        [
            (DEPTH, self.level.to_string()),
            (CEILING, self.ceiling.to_string()),
        ]
    }
}

/// One variable, where it is set and reads as a number.
fn read(name: &str) -> Option<u32> {
    env::var(name).ok()?.trim().parse().ok()
}
